lib.systems.doubles: add missing MIPS linux doubles

alyssais opened the pull request:
###### Description of changes

We already have examples for all of these, but since we didn't
actually recognise the doubles, it wasn't possible to build any
packages for them without setting allowUnsupportedSystem.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
lib.systems.doubles: add big-endian MIPS linux doubles

We already have examples for these, but since we didn't actually
recognise the doubles, it wasn't possible to build any packages for
them without setting allowUnsupportedSystem.

SuperSandro2000 commented on lib/tests/systems.nix:
```diff
@@ -34,7 +34,7 @@ with lib.systems.doubles; lib.runTests {
   testgnu = mseteq gnu (linux /* ++ kfreebsd ++ ... */);
   testillumos = mseteq illumos [ "x86_64-solaris" ];
-  testlinux = mseteq linux [ "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux" "armv7l-linux" "i686-linux" "mips64el-linux" "mipsel-linux" "riscv32-linux" "riscv64-linux" "x86_64-linux" "powerpc64-linux" "powerpc64le-linux" "m68k-linux" "s390-linux" "s390x-linux" "microblaze-linux" "microblazeel-linux" "loongarch64-linux" ];
+  testlinux = mseteq linux [ "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux" "armv7l-linux" "i686-linux" "loongarch64-linux" "m68k-linux" "microblaze-linux" "microblazeel-linux" "mips-linux" "mips64-linux" "mips64el-linux" "mipsel-linux" "mipsisa32r6-linux" "mipsisa32r6el-linux" "mipsisa64r6-linux" "mipsisa64r6el-linux" "powerpc64-linux" "powerpc64le-linux" "riscv32-linux" "riscv64 ...
```
```suggestion
  testlinux = mseteq linux [
    "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux" "armv7l-linux"
    "loongarch64-linux" "m68k-linux" "microblaze-linux" "microblazeel-linux"
    "mips-linux" "mips64-linux" "mips64el-linux" "mipsel-linux" "mipsisa32r6-linux" "mipsisa32r6el-linux" "mipsisa64r6-linux" "mipsisa64r6el-linux"
    "powerpc64-linux" "powerpc64le-linux" "riscv32-linux" "riscv64-linux" "s390-linux" "s390x-linux"
    "i686-linux" "x86_64-linux"
  ];
```
that line is a bit very long now. Maybe we can weakly group some things together? Feel free to shuffle things around.

SuperSandro2000 commented on lib/systems/doubles.nix:
```diff
@@ -27,7 +27,9 @@ let
     "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux"
     "armv7l-linux" "i686-linux" "loongarch64-linux" "m68k-linux" "microblaze-linux"
-    "microblazeel-linux" "mipsel-linux" "mips64el-linux" "powerpc64-linux"
+    "microblazeel-linux" "mips-linux" "mips64-linux" "mips64el-linux"
+    "mipsel-linux" "mipsisa32r6-linux" "mipsisa32r6el-linux"
+    "mipsisa64r6-linux" "mipsisa64r6el-linux" "powerpc64-linux"
```
We could do the same here and group arm, mipsel, power/s390, riscv and others or so but it is really just formatting which is not really important.

amjoseph-nixpkgs reviewed (commented):
LGTM if you drop the `glibc.meta.broken=true` or wait for me to diagnose it.

Might want to specify which doubles were missing: big-endian and `isa{32,64}r6`.

IMHO it was a mistake for me to add the `isa{32,64}r6` triples in the first place.  I was just imitating what Debian does.  But those triples differ from the "vanilla" triples only in the addition of a `-march=` flag.  If you'd prefer to delete all of the `isa{32,64}r6` stuff from `lib/systems/` instead I think that would probably be a better way to go.  Obviously the big-endian additions to `lib/systems/doubles.nix` in this PR are a good thing and should be kept.

alyssais committed:
lib.systems: remove mipsisa(32|64)r6 triples

These arc the same as the normal triples apart for a difference in
-march, so there's no need for them to be separate triples.

alyssais commented on pkgs/development/libraries/glibc/common.nix:
```diff
@@ -268,5 +268,7 @@ stdenv.mkDerivation ({
     maintainers = with maintainers; [ eelco ma27 ];
     platforms = platforms.linux;
+
+    broken = stdenv.hostPlatform.parsed.cpu.name == "mipsel";
```
Checked again, and this time it builds fine.  Not sure what happened before… I've dropped that commit.

alyssais commented:
> If you'd prefer to delete all of the `isa{32,64}r6` stuff from `lib/systems/` instead I think that would probably be a better way to go.

Done.

amjoseph-nixpkgs reviewed (approved):
Thanks for doing this!!
